org/mozilla/jss/provider/javax/net/JSSContext.java

```diff
@@ -0,0 +1,116 @@
+package org.mozilla.jss.provider.javax.net;
+
+import java.security.*;
+import java.util.ArrayList;
+
+import javax.net.ssl.*;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.mozilla.jss.CryptoManager;
+import org.mozilla.jss.provider.javax.crypto.JSSKeyManager;
+import org.mozilla.jss.provider.javax.crypto.JSSTrustManager;
+import org.mozilla.jss.pkcs11.PK11Cert;
+import org.mozilla.jss.pkcs11.PK11PrivKey;
+import org.mozilla.jss.ssl.javax.JSSEngine;
+import org.mozilla.jss.ssl.javax.JSSEngineReferenceImpl;
+import org.mozilla.jss.ssl.SSLVersion;
+
+public class JSSContext extends SSLContextSpi {
+    public static Logger logger = LoggerFactory.getLogger(JSSContext.class);
+
+    JSSKeyManager key_manager;
+    X509TrustManager trust_managers[];
+
+    SSLVersion protocol_version;
+
+    public void engineInit(KeyManager[] kms, TrustManager[] tms, SecureRandom sr) throws KeyManagementException {
+        logger.debug("JSSContext: engineInit(" + kms + ", " + tms + ", " + sr + ")");
+
+        if (kms != null) {
+            for (KeyManager km : kms) {
+                if (km instanceof JSSKeyManager) {
+                    key_manager = (JSSKeyManager) km;
+                    break;
+                }
+            }
+        }
+
+        if (tms != null) {
+            ArrayList<X509TrustManager> xtms = new ArrayList<X509TrustManager>();
+            for (TrustManager tm : tms) {
+                if (tm instanceof X509TrustManager) {
+                    xtms.add((X509TrustManager) tm);
+                }
+            }
+
+            trust_managers = xtms.toArray(new X509TrustManager[xtms.size()]);
+        }
```

Comment:
Eventually we're not going to care about the KeyManagers or TrustManagers as much, and push them all directly to our SSLEngine. I still need to verify what the logic in other SSLEngines look like (re: multiple TrustManagers and verifying certificates), but I largely think we can drop most of the special-casing around JSSTrustManagers / JSSKeyManagers and treat them all as X509TrustManagers/X509KeyManagers.

The only special casing that will remain is a special "indicator" JSSNativeTrustManager here to use the native cert-checking code we had in `SSLSocket` (the only way that does OCSP checking for now).

Reply:
The other comment I have here is that I don't think there's really any value to extending `TrustManager`: it is an [empty interface](https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/TrustManager.html). The only known derived interface here is [`X509TrustManager`](https://docs.oracle.com/javase/8/docs/api/javax/net/ssl/X509TrustManager.html). So if we're passed a set of `TrustManager`s with no `X509TrustManager`s, I'm not really sure what we're passed. :-)

This is mostly to ensure we have fewer conversions and filter out mostly useless junk. I think our library should default to something useful if at all possible, and fail otherwise. SNI likely will be a null `KeyManager` passed in (or perhaps a special placeholder `KeyManager`) as we likely want to query the keys directly from the NSS DB. But I still need to read up more on configuring NSS's SSL interfaces and server cert selection.